File: apps/controller/android/ios/App/App/ControllerCredentialSession.swift
import Foundation

enum ControllerCredentialKind: String {
    case longTerm
    case temporary

    init(wire: String) throws {
        guard let value = ControllerCredentialKind(rawValue: wire) else {
            throw NSError(domain: "WhaleMaidCredential", code: 1, userInfo: [NSLocalizedDescriptionKey: "credentialKind must be longTerm or temporary"])
        }
        self = value
    }
}

struct ControllerCredentialSession {
    private(set) var server = ""
    private(set) var deviceId = ""
    private(set) var password = ""
    private(set) var sessionToken = ""
    private(set) var credentialKind: ControllerCredentialKind = .longTerm

    var connected: Bool { !server.isEmpty && !deviceId.isEmpty && !sessionToken.isEmpty }

    mutating func commit(server: String, deviceId: String, password: String, kind: ControllerCredentialKind, token: String) throws {
        guard !token.isEmpty else { throw NSError(domain: "WhaleMaidCredential", code: 2, userInfo: [NSLocalizedDescriptionKey: "relay response missing sessionToken"]) }
        self.server = server
        self.deviceId = deviceId
        self.password = kind == .longTerm ? password : ""
        self.credentialKind = kind
        self.sessionToken = token
    }

    mutating func updateToken(_ token: String) throws {
        guard !token.isEmpty else { throw NSError(domain: "WhaleMaidCredential", code: 4, userInfo: [NSLocalizedDescriptionKey: "relay response missing sessionToken"]) }
        sessionToken = token
    }

    func payload(useToken: Bool) throws -> [String: Any] {
        if useToken { return ["deviceId": deviceId, "sessionToken": sessionToken] }
        guard credentialKind == .longTerm, !password.isEmpty else {
            throw NSError(domain: "WhaleMaidCredential", code: 3, userInfo: [NSLocalizedDescriptionKey: "temporary credential cannot fall back to password"])
        }
        return ["deviceId": deviceId, "password": password, "credentialKind": ControllerCredentialKind.longTerm.rawValue]
    }

    func canFallbackPassword(status: Int) -> Bool {
        status == 401 && credentialKind == .longTerm && !password.isEmpty
    }
}

File: apps/controller/android/android/app/src/main/java/com/whalemaid/controller/ControllerCredentialSession.kt
package com.whalemaid.controller

import org.json.JSONObject

enum class CredentialKind(val wire: String) {
    LONG_TERM("longTerm"),
    TEMPORARY("temporary");

    companion object {
        fun fromWire(value: String): CredentialKind = entries.firstOrNull { it.wire == value }
            ?: throw IllegalArgumentException("credentialKind must be longTerm or temporary")
    }
}

class ControllerCredentialSession {
    @Volatile var server: String = ""
        private set
    @Volatile var deviceId: String = ""
        private set
    @Volatile var password: String = ""
        private set
    @Volatile var authToken: String = ""
    @Volatile var credentialKind: CredentialKind = CredentialKind.LONG_TERM
        private set

    val connected: Boolean
        get() = server.isNotEmpty() && deviceId.isNotEmpty() && authToken.isNotEmpty()

    fun commit(
        server: String,
        deviceId: String,
        password: String,
        credentialKind: CredentialKind,
        authToken: String,
    ) {
        require(authToken.isNotEmpty()) { "relay response missing sessionToken" }
        this.server = server
        this.deviceId = deviceId
        this.password = if (credentialKind == CredentialKind.LONG_TERM) password else ""
        this.credentialKind = credentialKind
        this.authToken = authToken
    }

    fun payload(useToken: Boolean): JSONObject {
        val value = JSONObject().put("deviceId", deviceId)
        if (useToken) return value.put("sessionToken", authToken)
        check(credentialKind == CredentialKind.LONG_TERM && password.isNotEmpty()) {
            "temporary credential cannot fall back to password"
        }
        return value
            .put("password", password)
            .put("credentialKind", CredentialKind.LONG_TERM.wire)
    }

    fun canFallbackPassword(status: Int): Boolean =
        status == 401 && credentialKind == CredentialKind.LONG_TERM && password.isNotEmpty()
}
